lib/ansible/playbook/handler.py

```diff
@@ -49,6 +49,19 @@ def flag_for_host(self, host):
     def has_triggered(self, host):
         return host in self._flagged_hosts
 
+    def in_scope(self, templar, handler_vars):
+        '''
+        When handler is inside a role, this checks if the role was actually
+        included by re-evaluating its when conditional. Used at handler search time
+        to skip non-included handlers.
+
+        https://github.com/ansible/ansible/issues/37512
+        '''
+        if self._role:
+            return self._role.evaluate_conditional(templar, handler_vars)
```

Comment:
How does this work with 'include_role'? it should always be true if role was included

Reply:
My understanding is that `in_scope` is always `True` for the "include_role roles" because in that case `role.when` is set to `[]` since `when` is tied to the `include_role` task and not the actual role itself. So it should work for `include_role` for "free".

As for "the fun", that will be issue for roles from `- roles:` in a playbook too right? Do we need to add a flag `included` that would be set only when the role is actually included and use that flag here in `in_scope` to check it instead of evaluating the conditional?

Reply:
> the 'fun' would be an include_role with a condition that is true when the role is included but vars change and would be false afterwards, though that might be the case on import also?

@bcoca I tried to create the above scenario like this:

playbook.yml
```
- hosts: localhost
  gather_facts: no
  pre_tasks:
    - set_fact:
         handlers_var: True
  roles:
    - name: handlers_cond
      when: handlers_var
```

roles/handlers_cond/tasks/main.yml
```
- set_fact:
    handlers_var: False

- name: Changed command
  command: /bin/true
  notify: handler execute
```

roles/handlers_cond/handlers/main.yml
```
- name: handler execute
  command: /bin/echo executed
```

But that results in (both with `devel` and this PR):
```
PLAY [localhost] ****************************************************************************************

TASK [set_fact] *****************************************************************************************
ok: [localhost]

TASK [handlers_cond : set_fact] *************************************************************************
ok: [localhost]

TASK [handlers_cond : Changed command] ******************************************************************
skipping: [localhost]

PLAY RECAP **********************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=1
```

The remainder of the role is skipped when the vars in the condition (that guards the whole role) are changed. So not sure this PR changes anything in that regard?

Reply:
Since role imports are non conditional, the `when` never applies to the inclusion of the role, the 'fun' part was dealing with `include_role` specifically, in which case that would be applied to the action of including the role itself.

Reply:
i'm puzzled on why we need to evaluate the conditional on the role, since the handler (as all tasks in role) should inherit any conditionals applied when importing, if that is not the case, then we should fix it to do so.

Reply:
We need to check if the handler was actually included, by evaluating the conditional on the role that includes the handler. Evaluating `hander.when` might result in skipping a handler that was included (which is the handler we want to use) but its conditional evaluates to `False` (so `handler.when == [role_cond_that_is_true, handler_cond_that_is_false]`).

Reply:
i'm not sure that makes sense, if handler was not included, it would not be available for notification already.  

Since it should already add the conditionals inherited (only on import or include + apply) you should not need to check any other conditions.

Reply:
> i'm not sure that makes sense, if handler was not included, it would not be available for notification already.

But that's a case only with `include_role` no? With `import_role` and `- roles:` it is available for notification with conditionals inherited from the role the handler is in.

> Since it should already add the conditionals inherited (only on import or include + apply) you should not need to check any other conditions.

As I am not sure we're on the same page here, let me describe this from scratch from my point of view. This PR deals with the following scenario (see attached integration test):
1. `role_master` is imported (via `roles:`)
2. `role_master`'s dependency is imported: `role_slave` with `when: False`, which means that all of its tasks and **handlers** are imported with `when: False` (inherited)
3. `role_master`'s tasks are executed; role `role_slave` is included again, this time via `include_role` (does not matter), which means that the same handler is saved in `notified_handlers` again, so now `notified_handlers` contains the same handler twice, one with `when: False` from `2.` and one with `when: []` from mentioned `include_role`
4. The handler is notified.

Now, since handlers look up works as first found, first used, the one with `when: False` (which is role's conditional inherited by the handler) is used and consequently NOT executed because of its conditional, which is incorrect as it is expected (or is it?) that the handler from `include_role` should be found instead.

And to fix this, we look at that role's conditional when searching for a handler and skip those that are evaluated to `False`.

So this deals with handlers being included first with `import_role` or `roles:` (NOT `include_role`) and having inherited "falsy" conditional. And then the same handler is included again and is shadowed by the first at the look up time.

Is that incorrect? If so, what would you suggest? Sorry in case you did already suggest an alternative; I don't follow.

Reply:
3. seems to be not working as expected.

It should not be the 'same handler', the new one should override the old one since they share the same name, so its not first found, but 'only found' which should be the 'last one with the same name'. 

If the behaviour is as described above, it seems we are not doing the update correctly and this is what we should fix. We probably avoiding a re-import of handlers not realizing the 'signature' of the role is different, death by optimization ...

Reply:
Soo, https://github.com/ansible/ansible/pull/49249 will fix the above scenario (which is #37512). However, with that merged, the inverted scenario will no longer work:
```
roles:
   - name: a
   - name: a
     when: False
```
if inside role `a` a handler is notified, the one with `when: False` will be looked up (since it's last loaded) and won't be executed. This PR fixes that, though, do we want that to work or are we ok with the last handler overwriting those before? From the user perspective, I'd say it is expected that the handler from the first role is used. Or maybe at least add a toggle?

Reply:
I would make sure we clarify the handler docs to make sure its clear:

a) handlers are play scoped
b) handlers with same name are overwritten, last one defined wins

If we were deduping i would have added a vvv('handler x is being overwritten') msg but since we are keeping the lists, just changing read order ... don't think we can do that.

Reply:
Also, remember we can notify handlers from almost anywhere:
``` yaml
- roles:
     -name: a
     - name: a
       when: False
- tasks:
     - action: 
        notify: handler_in_role_a
```